From 2611bfd9f7fdf845dfedf6dba79a920485c432f5 Mon Sep 17 00:00:00 2001
From: Markus Niebel <Markus.Niebel@tq-group.com>
Date: Tue, 20 Aug 2019 11:45:25 +0200
Subject: [PATCH] tqma8mx: tqma8mx-[1,2]gb-lpddr4_timing: update to rev.0005
 for HW rev.020x

add 3rd working point for HW REV.020x- now we have 3200,400,100 MTS

Signed-off-by: Markus Niebel <Markus.Niebel@tq-group.com>
---
 board/tqc/tqma8mx/tqma8mx-1gb-lpddr4_timing.c | 137 +++++++++++++++++-
 board/tqc/tqma8mx/tqma8mx-2gb-lpddr4_timing.c | 137 +++++++++++++++++-
 2 files changed, 266 insertions(+), 8 deletions(-)

diff --git a/board/tqc/tqma8mx/tqma8mx-1gb-lpddr4_timing.c b/board/tqc/tqma8mx/tqma8mx-1gb-lpddr4_timing.c
index 30e86156da..24eda4ec62 100644
--- a/board/tqc/tqma8mx/tqma8mx-1gb-lpddr4_timing.c
+++ b/board/tqc/tqma8mx/tqma8mx-1gb-lpddr4_timing.c
@@ -5,7 +5,7 @@
  *
  * Generated code from MX8M_DDR_tool
  * Align with uboot-imx_v2018.03_4.14.78_1.0.0_ga
- * REV.0004 1GB_K4F8E3S4HB-MFCJ03V-SEC813 (HW REV.020x)
+ * REV.0005 1GB_K4F8E3S4HB-MFCJ03V-SEC813 (HW REV.020x)
  * from MX8M_LPDDR4_RPA_v23.xlsx / MX8 DDR Stress Test V2.10
  */
 
@@ -82,6 +82,30 @@ static struct dram_cfg_param ddr_ddrc_cfg[] = {
 	{0x3d402190,0x3818200},
 	{0x3d402194,0x80303},
 	{0x3d4021b4,0x100},
+	{0x3d403020,0x1},
+	{0x3d403024,0x1f40},
+	{0x3d403050,0x20d040},
+	{0x3d403064,0x30005},
+	{0x3d4030dc,0x840000},
+	{0x3d4030e0,0x310000},
+	{0x3d4030e8,0x66004a},
+	{0x3d4030ec,0x16004a},
+	{0x3d403100,0xa010102},
+	{0x3d403104,0x30404},
+	{0x3d403108,0x203060b},
+	{0x3d40310c,0x505000},
+	{0x3d403110,0x2040202},
+	{0x3d403114,0x2030202},
+	{0x3d403118,0x1010004},
+	{0x3d40311c,0x301},
+	{0x3d403130,0x20300},
+	{0x3d403134,0xa100002},
+	{0x3d403138,0x5},
+	{0x3d403144,0x50003},
+	{0x3d403180,0xc0190004},
+	{0x3d403190,0x3818200},
+	{0x3d403194,0x80303},
+	{0x3d4031b4,0x100},
 	{0x3d400244,0x0},
 	{0x3d400250,0x29001505},
 	{0x3d400254,0x2c},
@@ -159,6 +183,14 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x11215f,0x1ff},
 	{0x11305f,0x1ff},
 	{0x11315f,0x1ff},
+	{0x21005f,0x1ff},
+	{0x21015f,0x1ff},
+	{0x21105f,0x1ff},
+	{0x21115f,0x1ff},
+	{0x21205f,0x1ff},
+	{0x21215f,0x1ff},
+	{0x21305f,0x1ff},
+	{0x21315f,0x1ff},
 	{0x55,0x1ff},
 	{0x1055,0x1ff},
 	{0x2055,0x1ff},
@@ -171,16 +203,22 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x9055,0x1ff},
 	{0x200c5,0x19},
 	{0x1200c5,0x7},
+	{0x2200c5,0x7},
 	{0x2002e,0x2},
 	{0x12002e,0x2},
+	{0x22002e,0x2},
 	{0x90204,0x0},
 	{0x190204,0x0},
+	{0x290204,0x0},
 	{0x20024,0x1ab},
 	{0x2003a,0x0},
 	{0x120024,0x1ab},
 	{0x2003a,0x0},
+	{0x220024,0x1ab},
+	{0x2003a,0x0},
 	{0x20056,0x3},
 	{0x120056,0xa},
+	{0x220056,0xa},
 	{0x1004d,0xe00},
 	{0x1014d,0xe00},
 	{0x1104d,0xe00},
@@ -197,6 +235,14 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x11214d,0xe00},
 	{0x11304d,0xe00},
 	{0x11314d,0xe00},
+	{0x21004d,0xe00},
+	{0x21014d,0xe00},
+	{0x21104d,0xe00},
+	{0x21114d,0xe00},
+	{0x21204d,0xe00},
+	{0x21214d,0xe00},
+	{0x21304d,0xe00},
+	{0x21314d,0xe00},
 	{0x10049,0xeba},
 	{0x10149,0xeba},
 	{0x11049,0xeba},
@@ -213,6 +259,14 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x112149,0xeba},
 	{0x113049,0xeba},
 	{0x113149,0xeba},
+	{0x210049,0xeba},
+	{0x210149,0xeba},
+	{0x211049,0xeba},
+	{0x211149,0xeba},
+	{0x212049,0xeba},
+	{0x212149,0xeba},
+	{0x213049,0xeba},
+	{0x213149,0xeba},
 	{0x43,0x63},
 	{0x1043,0x63},
 	{0x2043,0x63},
@@ -228,6 +282,7 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x20050,0x0},
 	{0x20008,0x320},
 	{0x120008,0x64},
+	{0x220008,0x19},
 	{0x20088,0x9},
 	{0x200b2,0xdc},
 	{0x10043,0x5a1},
@@ -247,25 +302,39 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x112143,0x5a1},
 	{0x113043,0x5a1},
 	{0x113143,0x5a1},
+	{0x2200b2,0xdc},
+	{0x210043,0x5a1},
+	{0x210143,0x5a1},
+	{0x211043,0x5a1},
+	{0x211143,0x5a1},
+	{0x212043,0x5a1},
+	{0x212143,0x5a1},
+	{0x213043,0x5a1},
+	{0x213143,0x5a1},
 	{0x200fa,0x1},
 	{0x1200fa,0x1},
+	{0x2200fa,0x1},
 	{0x20019,0x1},
 	{0x120019,0x1},
-	{0x200f0,0x60},
+	{0x220019,0x1},
+	{0x200f0,0x660},
 	{0x200f1,0x0},
 	{0x200f2,0x4444},
 	{0x200f3,0x8888},
-	{0x200f4,0x5565},
+	{0x200f4,0x5665},
 	{0x200f5,0x0},
 	{0x200f6,0x0},
 	{0x200f7,0xf000},
 	{0x20025,0x0},
 	{0x2002d,0x0},
 	{0x12002d,0x0},
+	{0x22002d,0x0},
 	{0x200c7,0x80},
 	{0x1200c7,0x80},
+	{0x2200c7,0x80},
 	{0x200ca,0x106},
 	{0x1200ca,0x106},
+	{0x2200ca,0x106},
 };
 
 /* ddr phy trained csr */
@@ -1071,6 +1140,47 @@ static struct dram_cfg_param ddr_fsp1_cfg[] = {
 };
 
 
+/* P2 message block paremeter for training firmware */
+static struct dram_cfg_param ddr_fsp2_cfg[] = {
+	{0xd0000, 0x0},
+	{0x54002,0x102},
+	{0x54003,0x64},
+	{0x54004,0x2},
+	{0x54005,0x2228},
+	{0x54006,0x11},
+	{0x54008,0x121f},
+	{0x54009,0xc8},
+	{0x5400b,0x2},
+	{0x5400d,0x100},
+	{0x54012,0x110},
+	{0x54019,0x84},
+	{0x5401a,0x31},
+	{0x5401b,0x4a66},
+	{0x5401c,0x4a08},
+	{0x5401e,0x16},
+	{0x5401f,0x84},
+	{0x54020,0x31},
+	{0x54021,0x4a66},
+	{0x54022,0x4a08},
+	{0x54024,0x16},
+	{0x5402b,0x1000},
+	{0x5402c,0x1},
+	{0x54032,0x8400},
+	{0x54033,0x3100},
+	{0x54034,0x6600},
+	{0x54035,0x84a},
+	{0x54036,0x4a},
+	{0x54037,0x1600},
+	{0x54038,0x8400},
+	{0x54039,0x3100},
+	{0x5403a,0x6600},
+	{0x5403b,0x84a},
+	{0x5403c,0x4a},
+	{0x5403d,0x1600},
+	{0xd0000, 0x1},
+};
+
+
 /* P0 2D message block paremeter for training firmware */
 static struct dram_cfg_param ddr_fsp0_2d_cfg[] = {
 	{0xd0000, 0x0},
@@ -1609,6 +1719,10 @@ static struct dram_cfg_param ddr_phy_pie[] = {
 	{0x12000c,0x19},
 	{0x12000d,0xfa},
 	{0x12000e,0x10},
+	{0x22000b,0x3},
+	{0x22000c,0x6},
+	{0x22000d,0x3e},
+	{0x22000e,0x10},
 	{0x9000c,0x0},
 	{0x9000d,0x173},
 	{0x9000e,0x60},
@@ -1621,6 +1735,8 @@ static struct dram_cfg_param ddr_phy_pie[] = {
 	{0x20011,0x3},
 	{0x120010,0x5a},
 	{0x120011,0x3},
+	{0x220010,0x5a},
+	{0x220011,0x3},
 	{0x40080,0xe0},
 	{0x40081,0x12},
 	{0x40082,0xe0},
@@ -1633,6 +1749,12 @@ static struct dram_cfg_param ddr_phy_pie[] = {
 	{0x140083,0x12},
 	{0x140084,0xe0},
 	{0x140085,0x12},
+	{0x240080,0xe0},
+	{0x240081,0x12},
+	{0x240082,0xe0},
+	{0x240083,0x12},
+	{0x240084,0xe0},
+	{0x240085,0x12},
 	{0x400fd,0xf},
 	{0x10011,0x1},
 	{0x10012,0x1},
@@ -1710,6 +1832,13 @@ static struct dram_fsp_msg ddr_dram_fsp_msg[] = {
 		.fsp_cfg = ddr_fsp1_cfg,
 		.fsp_cfg_num = ARRAY_SIZE(ddr_fsp1_cfg),
 	},
+	{
+		/* P2 100mts 1D */
+		.drate = 100,
+		.fw_type = FW_1D_IMAGE,
+		.fsp_cfg = ddr_fsp2_cfg,
+		.fsp_cfg_num = ARRAY_SIZE(ddr_fsp2_cfg),
+	},
 	{
 		/* P0 3200mts 2D */
 		.drate = 3200,
@@ -1731,6 +1860,6 @@ struct dram_timing_info tqma8mx_1gb_dram_timing = {
 	.ddrphy_trained_csr_num = ARRAY_SIZE(ddr_ddrphy_trained_csr),
 	.ddrphy_pie = ddr_phy_pie,
 	.ddrphy_pie_num = ARRAY_SIZE(ddr_phy_pie),
-	.fsp_table = { 3200, 400, },
+	.fsp_table = { 3200, 400, 100, },
 };
 
diff --git a/board/tqc/tqma8mx/tqma8mx-2gb-lpddr4_timing.c b/board/tqc/tqma8mx/tqma8mx-2gb-lpddr4_timing.c
index 03536e292a..159dc7c6b0 100644
--- a/board/tqc/tqma8mx/tqma8mx-2gb-lpddr4_timing.c
+++ b/board/tqc/tqma8mx/tqma8mx-2gb-lpddr4_timing.c
@@ -5,7 +5,7 @@
  *
  * Generated code from MX8M_DDR_tool
  * Align with uboot-imx_v2018.03_4.14.78_1.0.0_ga
- * REV.0004 2GB_K4F6E3S4HM-GFCL03V-SEC804 (HW REV.020x)
+ * REV.0005 2GB_K4F6E3S4HM-GFCL03V-SEC804 (HW REV.020x)
  * from MX8M_LPDDR4_RPA_v23.xlsx / MX8 DDR Stress Test V2.10
  */
 
@@ -82,6 +82,30 @@ static struct dram_cfg_param ddr_ddrc_cfg[] = {
 	{0x3d402190,0x3818200},
 	{0x3d402194,0x80303},
 	{0x3d4021b4,0x100},
+	{0x3d403020,0x1},
+	{0x3d403024,0x1f40},
+	{0x3d403050,0x20d040},
+	{0x3d403064,0x30007},
+	{0x3d4030dc,0x840000},
+	{0x3d4030e0,0x310000},
+	{0x3d4030e8,0x66004a},
+	{0x3d4030ec,0x16004a},
+	{0x3d403100,0xa010102},
+	{0x3d403104,0x30404},
+	{0x3d403108,0x203060b},
+	{0x3d40310c,0x505000},
+	{0x3d403110,0x2040202},
+	{0x3d403114,0x2030202},
+	{0x3d403118,0x1010004},
+	{0x3d40311c,0x301},
+	{0x3d403130,0x20300},
+	{0x3d403134,0xa100002},
+	{0x3d403138,0x8},
+	{0x3d403144,0x50003},
+	{0x3d403180,0xc0190004},
+	{0x3d403190,0x3818200},
+	{0x3d403194,0x80303},
+	{0x3d4031b4,0x100},
 	{0x3d400244,0x0},
 	{0x3d400250,0x29001505},
 	{0x3d400254,0x2c},
@@ -159,6 +183,14 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x11215f,0x1ff},
 	{0x11305f,0x1ff},
 	{0x11315f,0x1ff},
+	{0x21005f,0x1ff},
+	{0x21015f,0x1ff},
+	{0x21105f,0x1ff},
+	{0x21115f,0x1ff},
+	{0x21205f,0x1ff},
+	{0x21215f,0x1ff},
+	{0x21305f,0x1ff},
+	{0x21315f,0x1ff},
 	{0x55,0x1ff},
 	{0x1055,0x1ff},
 	{0x2055,0x1ff},
@@ -171,16 +203,22 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x9055,0x1ff},
 	{0x200c5,0x19},
 	{0x1200c5,0x7},
+	{0x2200c5,0x7},
 	{0x2002e,0x2},
 	{0x12002e,0x2},
+	{0x22002e,0x2},
 	{0x90204,0x0},
 	{0x190204,0x0},
+	{0x290204,0x0},
 	{0x20024,0x1ab},
 	{0x2003a,0x0},
 	{0x120024,0x1ab},
 	{0x2003a,0x0},
+	{0x220024,0x1ab},
+	{0x2003a,0x0},
 	{0x20056,0x3},
 	{0x120056,0xa},
+	{0x220056,0xa},
 	{0x1004d,0xe00},
 	{0x1014d,0xe00},
 	{0x1104d,0xe00},
@@ -197,6 +235,14 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x11214d,0xe00},
 	{0x11304d,0xe00},
 	{0x11314d,0xe00},
+	{0x21004d,0xe00},
+	{0x21014d,0xe00},
+	{0x21104d,0xe00},
+	{0x21114d,0xe00},
+	{0x21204d,0xe00},
+	{0x21214d,0xe00},
+	{0x21304d,0xe00},
+	{0x21314d,0xe00},
 	{0x10049,0xeba},
 	{0x10149,0xeba},
 	{0x11049,0xeba},
@@ -213,6 +259,14 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x112149,0xeba},
 	{0x113049,0xeba},
 	{0x113149,0xeba},
+	{0x210049,0xeba},
+	{0x210149,0xeba},
+	{0x211049,0xeba},
+	{0x211149,0xeba},
+	{0x212049,0xeba},
+	{0x212149,0xeba},
+	{0x213049,0xeba},
+	{0x213149,0xeba},
 	{0x43,0x63},
 	{0x1043,0x63},
 	{0x2043,0x63},
@@ -228,6 +282,7 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x20050,0x0},
 	{0x20008,0x320},
 	{0x120008,0x64},
+	{0x220008,0x19},
 	{0x20088,0x9},
 	{0x200b2,0xdc},
 	{0x10043,0x5a1},
@@ -247,25 +302,39 @@ static struct dram_cfg_param ddr_ddrphy_cfg[] = {
 	{0x112143,0x5a1},
 	{0x113043,0x5a1},
 	{0x113143,0x5a1},
+	{0x2200b2,0xdc},
+	{0x210043,0x5a1},
+	{0x210143,0x5a1},
+	{0x211043,0x5a1},
+	{0x211143,0x5a1},
+	{0x212043,0x5a1},
+	{0x212143,0x5a1},
+	{0x213043,0x5a1},
+	{0x213143,0x5a1},
 	{0x200fa,0x1},
 	{0x1200fa,0x1},
+	{0x2200fa,0x1},
 	{0x20019,0x1},
 	{0x120019,0x1},
-	{0x200f0,0x60},
+	{0x220019,0x1},
+	{0x200f0,0x660},
 	{0x200f1,0x0},
 	{0x200f2,0x4444},
 	{0x200f3,0x8888},
-	{0x200f4,0x5565},
+	{0x200f4,0x5665},
 	{0x200f5,0x0},
 	{0x200f6,0x0},
 	{0x200f7,0xf000},
 	{0x20025,0x0},
 	{0x2002d,0x0},
 	{0x12002d,0x0},
+	{0x22002d,0x0},
 	{0x200c7,0x80},
 	{0x1200c7,0x80},
+	{0x2200c7,0x80},
 	{0x200ca,0x106},
 	{0x1200ca,0x106},
+	{0x2200ca,0x106},
 };
 
 /* ddr phy trained csr */
@@ -1071,6 +1140,47 @@ static struct dram_cfg_param ddr_fsp1_cfg[] = {
 };
 
 
+/* P2 message block paremeter for training firmware */
+static struct dram_cfg_param ddr_fsp2_cfg[] = {
+	{0xd0000, 0x0},
+	{0x54002,0x102},
+	{0x54003,0x64},
+	{0x54004,0x2},
+	{0x54005,0x2228},
+	{0x54006,0x11},
+	{0x54008,0x121f},
+	{0x54009,0xc8},
+	{0x5400b,0x2},
+	{0x5400d,0x100},
+	{0x54012,0x110},
+	{0x54019,0x84},
+	{0x5401a,0x31},
+	{0x5401b,0x4a66},
+	{0x5401c,0x4a08},
+	{0x5401e,0x16},
+	{0x5401f,0x84},
+	{0x54020,0x31},
+	{0x54021,0x4a66},
+	{0x54022,0x4a08},
+	{0x54024,0x16},
+	{0x5402b,0x1000},
+	{0x5402c,0x1},
+	{0x54032,0x8400},
+	{0x54033,0x3100},
+	{0x54034,0x6600},
+	{0x54035,0x84a},
+	{0x54036,0x4a},
+	{0x54037,0x1600},
+	{0x54038,0x8400},
+	{0x54039,0x3100},
+	{0x5403a,0x6600},
+	{0x5403b,0x84a},
+	{0x5403c,0x4a},
+	{0x5403d,0x1600},
+	{0xd0000, 0x1},
+};
+
+
 /* P0 2D message block paremeter for training firmware */
 static struct dram_cfg_param ddr_fsp0_2d_cfg[] = {
 	{0xd0000, 0x0},
@@ -1609,6 +1719,10 @@ static struct dram_cfg_param ddr_phy_pie[] = {
 	{0x12000c,0x19},
 	{0x12000d,0xfa},
 	{0x12000e,0x10},
+	{0x22000b,0x3},
+	{0x22000c,0x6},
+	{0x22000d,0x3e},
+	{0x22000e,0x10},
 	{0x9000c,0x0},
 	{0x9000d,0x173},
 	{0x9000e,0x60},
@@ -1621,6 +1735,8 @@ static struct dram_cfg_param ddr_phy_pie[] = {
 	{0x20011,0x3},
 	{0x120010,0x5a},
 	{0x120011,0x3},
+	{0x220010,0x5a},
+	{0x220011,0x3},
 	{0x40080,0xe0},
 	{0x40081,0x12},
 	{0x40082,0xe0},
@@ -1633,6 +1749,12 @@ static struct dram_cfg_param ddr_phy_pie[] = {
 	{0x140083,0x12},
 	{0x140084,0xe0},
 	{0x140085,0x12},
+	{0x240080,0xe0},
+	{0x240081,0x12},
+	{0x240082,0xe0},
+	{0x240083,0x12},
+	{0x240084,0xe0},
+	{0x240085,0x12},
 	{0x400fd,0xf},
 	{0x10011,0x1},
 	{0x10012,0x1},
@@ -1710,6 +1832,13 @@ static struct dram_fsp_msg ddr_dram_fsp_msg[] = {
 		.fsp_cfg = ddr_fsp1_cfg,
 		.fsp_cfg_num = ARRAY_SIZE(ddr_fsp1_cfg),
 	},
+	{
+		/* P2 100mts 1D */
+		.drate = 100,
+		.fw_type = FW_1D_IMAGE,
+		.fsp_cfg = ddr_fsp2_cfg,
+		.fsp_cfg_num = ARRAY_SIZE(ddr_fsp2_cfg),
+	},
 	{
 		/* P0 3200mts 2D */
 		.drate = 3200,
@@ -1731,6 +1860,6 @@ struct dram_timing_info tqma8mx_2gb_dram_timing = {
 	.ddrphy_trained_csr_num = ARRAY_SIZE(ddr_ddrphy_trained_csr),
 	.ddrphy_pie = ddr_phy_pie,
 	.ddrphy_pie_num = ARRAY_SIZE(ddr_phy_pie),
-	.fsp_table = { 3200, 400, },
+	.fsp_table = { 3200, 400, 100, },
 };
 
